### PR TITLE
NEXT-00000 - Cleanup line items when recalculating

### PR DESCRIPTION
**Pull request by jasperP98**
### 1. Why is this change necessary?
When an order is created with conditional line items, and during an order edit one of those line items becomes irrelevant, it should be removed from the order. While the recalculation service will determine the correct total amount, these irrelevant line items are not being removed from the database when the order data is updated.

### 2. What does this change do, exactly?
This change ensures that conditional line items which no longer belong in an order during an edit are properly removed. While the recalculation service accurately calculates the correct total amount, previously these irrelevant line items were not being removed from the database when the order data was updated. This change addresses that issue by ensuring these items are correctly deleted from the order and the database during the update process.

### 3. Describe each step to reproduce the issue or behaviour.

1. Start by creating a new order that includes line items subject to certain conditions (e.g., discounts that apply only under specific criteria).
2. Apply a discount that has specific conditions, such as a minimum purchase amount or a specific combination of products.
3. Modify the order in such a way that one or more line items no longer meet the conditions for the discount (e.g., remove an item that was necessary to qualify for the discount).
4. Observe that the irrelevant line items, which no longer meet the discount conditions, have not been removed from the database, despite the recalculated order total reflecting their removal.

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: src/Core/Checkout/Cart/Order/RecalculationService.php [hunk 1]
```diff
@@ -88,7 +91,11 @@ public function recalculateOrder(string $orderId, Context $context, array $sales
         }
 
         // change scope to be able to write protected state fields of transactions and deliveries
-        $context->scope(Context::SYSTEM_SCOPE, function (Context $context) use ($orderData): void {
+        $context->scope(Context::SYSTEM_SCOPE, function (Context $context) use ($orderData, $order): void {
+            if (($lineItems = $order->getLineItems()) instanceof OrderLineItemCollection) {
+                $this->orderLineItemRepository->delete(array_values($lineItems->fmap(static fn (OrderLineItemEntity $lineItem) => ['id' => $lineItem->getId()])), $context);
```
Comment (keulinho) [inline src/Core/Checkout/Cart/Order/RecalculationService.php]: Can you maybe add a testcase for this one? I think it is quite critical and would help understand the problem.

And also i would suggest to not directly delete all lineitems from the original order, but rather perform a check which lineitems exist in the original order, that are not part of the recalculated order and delete only those, this makes it also more explicit why we need to the deletion

Comment (keulinho) [reply src/Core/Checkout/Cart/Order/RecalculationService.php]: also esp for huge orders this approach should have a performance benefit

Comment (jasperP98) [reply src/Core/Checkout/Cart/Order/RecalculationService.php]: Hi @keulinho i believe you are right! I adjusted the code and created a testcase

File: changelog/_unreleased/2024-06-06-cleanup-line-items-on-recalculation.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+title: Cleanup line items on recalculation
+issue: NEXT-00000
+author: Jasper Peeters
+author_email: jasper.peeters@meteor.be
+author_github: JasperP98
+---
+
+# Core
+
+* Cleanup line items on recalculation because otherwise the line items are not recalculated correctly
```
Comment (Bird87ZA) [inline changelog/_unreleased/2024-06-06-cleanup-line-items-on-recalculation.md]:
```suggestion
* Changed line item recalculation to fix a calculation error by deleting removed line items.
```

**Review by Bird87ZA — DISMISSED** [review]

File: changelog/_unreleased/2024-06-06-cleanup-line-items-on-recalculation.md (hunk 2 above)
Comment (jasperP98) [reply changelog/_unreleased/2024-06-06-cleanup-line-items-on-recalculation.md]: I updated the changelog

**Comment by Bird87ZA** [comment]
@jasperP98 one of your tests are failing. Soon as you have fixed it, I'll get this merged. :)

```
1) Shopware\Tests\Integration\Core\Checkout\Cart\Order\RecalculationServiceTest::testDeleteLineItemsAfterRecalculateOrderWitchInactiveProducts
Failed asserting that null is not null.

/home/runner/work/shopware/shopware/tests/integration/Core/Checkout/Cart/Order/RecalculationServiceTest.php:739
```

**Comment by jasperP98** [comment]
@Bird87ZA I tested it locally and it should be fine now but the checks keep failing. It seems because of something else, is that a possibility? I also fixed a small bug where when a line item is deleten because the product is inactive, its delivery position keeps its relation. This causes a 500 error in the recalculation process.

**Comment by Bird87ZA** [comment]
I'll import it and see if I can see what the issue is. :) I'll keep you updated.

**Comment by ghost** [comment]
Hello,

thank you for creating this pull request.
I have opened an issue on our Issue Tracker for you. See the issue link: https://issues.shopware.com/issues/NEXT-36704

Please use this issue to track the state of your pull request.

**Comment by Bird87ZA** [comment]
Your PR has been merged! Thanks for your contribution 💙

**Pull request closed without merging**